Issues/gvk28 Sustainable solution for displaying images with ImGui

mrumpelnik opened the pull request:
This addresses issue #28 and adds a script to modify the original imgui sources to add support for displaying images with ImGui.

_Gears-Vk's_ `imgui_manager` is responsible for creating and caching `DescriptorSets` and hands them out as `ImTextureID`  via `imgui_manager::get_or_create_texture(...)`.

Background functionality for transparent images is added via a new `ImageWithBg` function in `imgui_user.h` (not sure if `ImageWithBg` is the best name but making just another overload for `Image` may clash sometime in feature imgui updates)

johannesugb commented on external/tools/modify_imgui_for_gearsvk.ps1:
```diff
@@ -0,0 +1,292 @@
+	}
+
+	# Detect if in ImGui_ImplVulkan_SetupRenderState
+	if ($line -match "^static void ImGui_ImplVulkan_SetupRenderState\(") {
```
Maybe protect the regex against changes in white spaces?! That should work with `\s+`, meaning "any whitespace, and at least one of it" (I think, at least).

johannesugb replied:
Maybe the `static` does not have to be included in the regex...then we would be safe if the `static` declaration was ever removed. Do you think, this could happen? Would the location still be uniquely identifiable without the `static`?

johannesugb commented on external/tools/modify_imgui_for_gearsvk.ps1:
```diff
@@ -0,0 +1,292 @@
+	}
+
+	# Detect if in ImGui_ImplVulkan_RenderDrawData
+	if ($line -match "^void ImGui_ImplVulkan_RenderDrawData\(") {
```
Use `\s+` here, too!

johannesugb commented on external/tools/modify_imgui_for_gearsvk.ps1:
```diff
@@ -0,0 +1,292 @@
+	}
+
+	# Detect if in ImGui_ImplVulkan_CreateFontsTexture
+	if ($line -match "^bool ImGui_ImplVulkan_CreateFontsTexture\(") {
```
Use `\s+` here, too!

johannesugb commented on external/tools/modify_imgui_for_gearsvk.ps1:
```diff
@@ -0,0 +1,292 @@
+	}
+
+	# Detect if the function we have modified is left by relying on consistent code
+	# formatting, i.e. the closing '}' is at the beginning of the line
```
Hmm... that's a bit dangerous. Optimally, one would count opening and closing curly braces per line and make a decision based on that. At least we should add some white-space safety checks.

johannesugb commented on external/tools/modify_imgui_for_gearsvk.ps1:
```diff
@@ -0,0 +1,292 @@
+	# Remove font descriptor set binding in ImGui_ImplVulkan_SetupRenderState
+	if ($inSetupRenderState) {
+		# Remove unecessary descriptor binding
+		if (($line -match "VkDescriptorSet desc_set\[1\] = \{ g_DescriptorSet \};") -Or
```
`\s+` never hurts.

johannesugb commented on external/tools/modify_imgui_for_gearsvk.ps1:
```diff
@@ -0,0 +1,292 @@
+
+	# Add descriptor set binding in ImGui_ImplVulkan_RenderDrawData
+	if ($inRenderDrawData) {
+		if ($line -match "vkCmdSetScissor\(command_buffer, 0, 1, &scissor\);") {
```
`\s+` never hurts. Hm.. what if the variable names ever change? Maybe it would be better to match arbitrary characters; The commas and braces should probably identify the correct line regardless!

johannesugb commented on external/tools/modify_imgui_for_gearsvk.ps1:
```diff
@@ -0,0 +1,292 @@
+
+	# Change fonts texture id assignment to descriptor in ImGui_ImplVulkan_CreateFontsTexture
+	if ($inCreateFontsTexture) {
+		if ($line -match "io.Fonts->SetTexID\(\(ImTextureID\)\(intptr_t\)g_FontImage\);") {
```
Maybe also here some safety-regex-measures could be applied. E.g. `\s*` means: "whitespaces, either zero or multiple".

mrumpelnik commented:
Updated to latest ImGui release (v1.83) and changed direct `ImDrawCmd::TextureId` access in RenderDrawData to the newly introduced function `ImDrawCmd::GetTexID()` since according to [imgui_v1.83](https://github.com/ocornut/imgui/releases/tag/v1.83) direct access is now obsolete and may break sometime in the future.

johannesugb commented on framework/include/imgui_manager.hpp:
```diff
@@ -42,6 +43,15 @@ namespace gvk
+		 *	@param[in] aImageSampler		An image sampler, that should be rendered with ImGui
+		 *  @param[out] ImTextureID			ImGui identifier for textures
+		 */
+		ImTextureID get_or_create_texture(avk::image_sampler& aImageSampler);
```
For passing resources, please use either `avk::resource_reference<const T>` or `avk::resource_reference<T>`  in that case (see _Auto-Vk_'s [Resource Management](https://github.com/cg-tuwien/Auto-Vk#resource-management)).

However, I think that in this case, no concrete `image_sampler` instance is even required... could that be? 
If I see it right, you could just use the following overload:
```
template <typename T>
binding_data descriptor_binding(uint32_t aSet, uint32_t aBinding, uint32_t aCount, shader_type aShaderStages = shader_type::all)
```
and call it with `T=avk::image_sampler_t`.

mrumpelnik replied:
Not sure I can follow: "no concrete image_sampler instance is even required". `imgui_manager::get_or_create_texture` currently creates one `DescriptorSet` per user texture (The sampler in this case). If we remove the sampler argument and simply use `avk::descriptor_binding<avk::image_sampler_t>(0u, 0u, 1u, avk::shader_type::fragment)`, the function creates a `DescriptorSet` once (without an user texture specified in the layout) and always returns that same one? 

Then how would code with multiple user textures, from i.e. ` compute_image_processing` such as 
```
ImTextureID inputTexId = imguiManager->get_or_create_texture(mInputImageAndSampler);
ImGui::Image(inputTexId, ...)
ImTextureID targetTexId = imguiManager->get_or_create_texture(mTargetImageAndSampler);
ImGui::Image(targetTexId, ...)
```
work if we don't specify to which `Sampler` we want the `DescriptorSet`? Or did I completely missunderstood your intention?

johannesugb replied:
Okay, sorry... I think, I confused myself about descriptor set layouts vs. concrete descriptor set instances - in this case, a concrete descriptor set needs to be created which, of course, need the concrete image handle! So, all good, I guess. 

But I think we should rename `get_or_create_texture` to `get_or_create_texture_id` or maybe to `get_or_create_texture_descriptor` because otherwise there could be misunderstandings in what the method does. "Creating a texture" sounds like allocating an image resource, which it does not do.  I think, I prefer a name which includes "descriptor".

mrumpelnik replied:
Hmm.. I agree `get_or_create_texture` seems confusing because you are not creating a texture. I think your suggestion `get_or_create_texture_descriptor` sounds good as it makes clear, what the function really does.

johannesugb commented on framework/src/imgui_manager.cpp:
```diff
@@ -6,6 +6,23 @@
 #include <GLFW/glfw3native.h>   // for glfwGetWin32Window
 #include <imgui_internal.h>
 
+//// Custom specialization of std::hash for DescriptorKey can be injected in namespace std
+//// adopted from https://en.cppreference.com/w/cpp/utility/hash to use hash_combine.
+//namespace std {
+//	template <>
+//	struct hash<DescriptorKey>
+//	{
+//		inline std::size_t operator()(const DescriptorKey& k) const
+//		{
+//			size_t seed = 0;
+//			hash_combine(seed, k.sampler);
+//			hash_combine(seed, k.image_view);
+//			hash_combine(seed, k.image_layout);
+//			return seed;
+//		}
+//	};
+//}
+
```
delete?

johannesugb commented on framework/src/imgui_manager.cpp:
```diff
@@ -37,7 +54,8 @@ namespace gvk
 
-		const uint32_t magicImguiFactor = 1000;
+		//const uint32_t magicImguiFactor = 1000;
+		const uint32_t magicImguiFactor = 1;
```
That's very cool :) I've still got to test this on AMD for safety, but Intel and Nvidia seem to be totally fine.

johannesugb commented on framework/src/imgui_manager.cpp:
```diff
@@ -392,4 +413,17 @@ namespace gvk
 		avk::assign_and_lifetime_handle_previous(mClearRenderpass, std::move(newClearRenderpass), lifetimeHandlerLambda);
 	}
+
+	ImTextureID imgui_manager::get_or_create_texture(avk::image_sampler& aImageSampler)
```
See comment above about whether the parameter is even required.

johannesugb commented on framework/src/imgui_manager.cpp:
```diff
@@ -392,4 +413,17 @@ namespace gvk
+	ImTextureID imgui_manager::get_or_create_texture(avk::image_sampler& aImageSampler)
+	{
+		std::vector<avk::descriptor_set> sets = mImTextureDescriptorCache.get_or_create_descriptor_sets({
+			avk::descriptor_binding(0, 0, aImageSampler, avk::shader_type::fragment)
```
```suggestion
			avk::descriptor_binding<avk::image_sampler_t>(0u, 0u, 1u, avk::shader_type::fragment)
```
Maybe that ^ suffices.

johannesugb replied:
Just ignore this ;)

mrumpelnik commented on external/tools/modify_imgui_for_gearsvk.ps1:
```diff
@@ -0,0 +1,331 @@
+	}
+
+	# Add user texture binding to implemented features list
+	if ($line -match "//\s*Missing\s*features:") {
```
Powershells `-match` is case insensitive by default but we could use `-imatch` to make the intention more explicit

johannesugb replied:
Oh, really? Didn't know that (and it appears a bit strange to me ^^) 
Yes, `-imatch` would be cool, because then there can be no misunderstandings and it isn't at risk of changing its behavior, which `-match` might be (if the user changed the default behavior?).

mrumpelnik replied:
Yes, I discovered and googled it during regex testing because strings matched, when they shouldn't ^^
According to [About Comparison Operator](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_comparison_operators?view=powershell-7.1) all comparison operators are case-insensitive by default and I'm not sure if you can change this behaviour but you are right and I will change everything to `-imatch` because this default behaviour is a bit confusing for people who are not working with Powershell every day

mrumpelnik commented on external/tools/modify_imgui_for_gearsvk.ps1:
```diff
@@ -0,0 +1,331 @@
+# - User texture binding as missing feature comment at the top:
+#	//  [ ] Renderer: User texture binding
+#
+# - The unecessary font texture descriptor binding:
```
Nothing functionally. Just removes the comment at the top, that user texture binding is missing because after our changes, it isn't missing anymore

johannesugb replied:
Okay.

johannesugb commented on external/tools/README.md:
```diff
@@ -0,0 +1,9 @@
+# How-to update imgui for Gears-Vk
```
```suggestion
# How-To Update ImGui for Gears-Vk
```

johannesugb commented on external/tools/README.md:
```diff
@@ -0,0 +1,9 @@
+# How-to update imgui for Gears-Vk
+- Checkout [imgui](https://github.com/ocornut/imgui)
```
```suggestion
- Checkout [ImGui](https://github.com/ocornut/imgui)
```

johannesugb commented on external/tools/README.md:
```diff
@@ -0,0 +1,9 @@
+- Copy `imgui/backends/imgui_impl_glfw.cpp` and `imgui/backends/imgui_impl_vulkan.cpp` into `external/universal/src`
+- Copy every `.h` file from `imgui/` into `external/universal/include`
+- Copy `imgui/backends/imgui_impl_glfw.h` and `imgui/backends/imgui_impl_vulkan.h` into `external/universal/include`
+- Open Powershell, change directory to `external/tools` and execute the script `modify_imgui_for_gearsvk.ps1`
```
```suggestion
- Open PowerShell, change directory to `external/tools/` and execute the script `modify_imgui_for_gearsvk.ps1`
```

johannesugb replied:
```suggestion
- Open Powershell, change directory to `external/tools/` and execute the script `modify_imgui_for_gearsvk.ps1`
```

johannesugb commented on external/tools/README.md:
```diff
@@ -0,0 +1,9 @@
+# How-to update imgui for Gears-Vk
+- Checkout [imgui](https://github.com/ocornut/imgui)
+- Copy every `.cpp` file from `imgui/` into `external/universal/src`
+- Copy `imgui/backends/imgui_impl_glfw.cpp` and `imgui/backends/imgui_impl_vulkan.cpp` into `external/universal/src`
```
```suggestion
- Copy `imgui/backends/imgui_impl_glfw.cpp` and `imgui/backends/imgui_impl_vulkan.cpp` into `external/universal/src/`
```

johannesugb commented on external/tools/README.md:
```diff
@@ -0,0 +1,9 @@
+- Checkout [imgui](https://github.com/ocornut/imgui)
+- Copy every `.cpp` file from `imgui/` into `external/universal/src`
+- Copy `imgui/backends/imgui_impl_glfw.cpp` and `imgui/backends/imgui_impl_vulkan.cpp` into `external/universal/src`
+- Copy every `.h` file from `imgui/` into `external/universal/include`
```
```suggestion
- Copy every `.h` file from `imgui/` into `external/universal/include/`
```

johannesugb commented on external/tools/README.md:
```diff
@@ -0,0 +1,9 @@
+- Copy every `.h` file from `imgui/` into `external/universal/include`
+- Copy `imgui/backends/imgui_impl_glfw.h` and `imgui/backends/imgui_impl_vulkan.h` into `external/universal/include`
+- Open Powershell, change directory to `external/tools` and execute the script `modify_imgui_for_gearsvk.ps1`
+- Check if every `.cpp` file in `external/universal/source` is included in the Visual Studio
```
```suggestion
- Check if every `.cpp` file in `external/universal/source/` is included in the Visual Studio framework library project `gears-vk`
```

johannesugb commented on external/tools/README.md:
```diff
@@ -0,0 +1,9 @@
+- Copy `imgui/backends/imgui_impl_glfw.h` and `imgui/backends/imgui_impl_vulkan.h` into `external/universal/include`
+- Open Powershell, change directory to `external/tools` and execute the script `modify_imgui_for_gearsvk.ps1`
+- Check if every `.cpp` file in `external/universal/source` is included in the Visual Studio
+  project
```
Add  comment about which console output is to be expected in the success case!
